src/Compilers/Core/Portable/StrongName/CryptoBlobParser.cs

```diff
@@ -165,106 +144,55 @@ private struct BlobHeader
-        [StructLayout(LayoutKind.Sequential)]
         private struct RsaPubKey
         {
             public UInt32 Magic; // Indicates RSA1 or RSA2
             public UInt32 BitLen; // Number of bits in the modulus. Must be multiple of 8.
             public UInt32 PubExp; // The public exponent
         }
+
+        private const int RsaPubKeySize = 12;
+
         private const UInt32 RSA1 = 0x31415352;
         private const UInt32 RSA2 = 0x32415352;
 
         // In wincrypt.h both public and private key blobs start with a
         // PUBLICKEYSTRUC and RSAPUBKEY and then start the key data
-        private unsafe static readonly int s_offsetToKeyData = sizeof(BlobHeader) + sizeof(RsaPubKey);
+        private const int s_offsetToKeyData = BlobHeaderSize + RsaPubKeySize;
 
-        private static ImmutableArray<byte> CreateSnPublicKeyBlob(BlobHeader header, RsaPubKey rsa, byte[] pubKeyData)
+        private static ImmutableArray<byte> CreateSnPublicKeyBlob(ref BlobHeader header, ref RsaPubKey rsa, byte[] pubKeyData)
         {
-            var snPubKey = new SnPublicKeyBlob()
-            {
-                SigAlgId = AlgorithmId.RsaSign,
-                HashAlgId = AlgorithmId.Sha,
-                PublicKeySize = (UInt32)(s_offsetToKeyData + pubKeyData.Length)
-            };
-
-            using (var ms = new MemoryStream(160))
-            using (var binaryWriter = new BinaryWriter(ms))
-            {
-                binaryWriter.Write(snPubKey.SigAlgId);
-                binaryWriter.Write(snPubKey.HashAlgId);
-                binaryWriter.Write(snPubKey.PublicKeySize);
+            var w = new BlobWriter(4 + 4 + 4 + s_offsetToKeyData + pubKeyData.Length);
+            w.WriteUInt32(AlgorithmId.RsaSign);
+            w.WriteUInt32(AlgorithmId.Sha);
+            w.WriteUInt32((uint)(s_offsetToKeyData + pubKeyData.Length));
 
-                binaryWriter.Write(header.Type);
-                binaryWriter.Write(header.Version);
-                binaryWriter.Write(header.Reserved);
-                binaryWriter.Write(header.AlgId);
+            w.WriteByte(header.Type);
+            w.WriteByte(header.Version);
+            w.WriteUInt16(header.Reserved);
+            w.WriteUInt32(header.AlgId);
 
-                binaryWriter.Write(rsa.Magic);
-                binaryWriter.Write(rsa.BitLen);
-                binaryWriter.Write(rsa.PubExp);
+            w.WriteUInt32(rsa.Magic);
+            w.WriteUInt32(rsa.BitLen);
+            w.WriteUInt32(rsa.PubExp);
 
-                binaryWriter.Write(pubKeyData);
+            w.WriteBytes(pubKeyData);
 
-                return ms.ToImmutable();
-            }
+            return w.ToImmutableArray();
         }
 
-        private unsafe static bool TryGetPublicKeyFromPublicKeyBlob(byte* blob, int blobLen, out ImmutableArray<byte> publicKey)
+        private static ImmutableArray<byte> TryGetPublicKeyFromPrivateKeyBlob(BlobHeader header, RsaPubKey rsa, byte[] pubKeyData)
```

Comment:
>TryGetPublicKeyFromPrivateKeyBlob [](start = 44, length = 33)

Has just one caller, I think it would be better to inline it and avoid reassigning the values of the header fields.